### ISSUE TITLE
Distributor instrumentation reports "success" to requests returning 4xx when called via HTTPgRPC

### ISSUE THREAD

**Issue by pracucci** [grafana/mimir#1830]
**Describe the bug**
In the distributor, the metric `cortex_request_duration_seconds` is counting 4xx as `status_code="success"` when the distributor `Push()` return an httpgrpc error with a 4xx code inside.

From a quick look, I believe the problem is in the httpgrpc server where only 5xx are converted into httpgprc errors:
https://github.com/weaveworks/common/blob/bd288de53d57de300fa286688ce2fc935687213f/httpgrpc/server/server.go#L67-L69

The grpc instrumentation then report everything which is not an error as `success`:
https://github.com/weaveworks/common/blob/bd288de53d57de300fa286688ce2fc935687213f/middleware/grpc_instrumentation.go#L14-L26

**Expected behavior**
To be reported as `status_code=4xx`.

See: https://github.com/cortexproject/cortex/issues/3949

_This issue affects Mimir too._

**Comment by pracucci** [comment]
The PR https://github.com/grafana/dskit/pull/457 proposes a change to return an `error` from the httpgrpc handler even when the status code is 4xx. This way the `error` will be captured by gRPC instrumentation and the status code tracked.

I agree with the strategy, but we need to check if there's any unintended side effect after this change. What I checked:

- Grafana Labs' cloud gateway:
  - gRPC proxy: correctly handle it
- Mimir:
  - `RemoteQuerier` calls `client.Handle()` in `RemoteQuerier.Read()` and `RemoteQuerier.sendRequest()`. I'm not convinced errors are handled the best way: we need to investigate it further and eventually improve it.

**Comment by duricanikolic** [comment]
## Analysis of usage of `httpgrpc`'s `server.Handle()` ##

### `frontendProcessor.runRequest()` and `schedluerProcessor.runRequest()` ###
As already pointed out by @pracucci, https://github.com/grafana/dskit/pull/457 will not change the current behaviour of the two functions. 

### `MultitenantAlertmanager.HandleRequest()` ###
This function is called by `pkg/alertmanager/Distributor.doRequest()`, which is called in 2 places:
- `pkg/alertmanager/Distributor.doUnary()`
- `pkg/alertmanager/Distributor.doQuorum()`

#### `pkg/alertmanager/Distributor.doUnary()` ####
The relevant part of this function is the following:
```
func (d *Distributor) doUnary(...) {
	...
	resp, err := d.doRequest(ctx, amDesc, req)
	if err != nil {
		respondFromError(err, w, logger)
		return
	}
	respondFromHTTPGRPCResponse(w, resp)
	...
}
func respondFromError(err error, w http.ResponseWriter, logger log.Logger) {
	httpResp, ok := httpgrpc.HTTPResponseFromError(errors.Cause(err))
	if !ok {
		level.Error(logger).Log("msg", "failed to process the request to the alertmanager", "err", err)
		http.Error(w, "Failed to process the request to the alertmanager", http.StatusInternalServerError)
		return
	}
	respondFromHTTPGRPCResponse(w, httpResp)
}

func respondFromHTTPGRPCResponse(w http.ResponseWriter, httpResp *httpgrpc.HTTPResponse) {
	for _, h := range httpResp.Headers {
		for _, v := range h.Values {
			w.Header().Add(h.Key, v)
		}
	}
	w.WriteHeader(int(httpResp.Code))
	w.Write(httpResp.Body) //nolint
}
```
Namely, in case of error returned by `d.doRequest()`, the error is first converted to an instance of `httpgrpc.HTTPResponse` (by `respondFromError()`), and then the latter is written to the writer (by `respondFromHTTPGRPCResponse()`). When no error is returned by `d.doRequest()`, the returned `httpgrpc.HTTPResponse` is written to the writer (by `respondFromHTTPGRPCResponse()`). 
Before https://github.com/grafana/dskit/pull/457, responses with `4xx` status code were not considered erroneous, and used to be returned as a `httpgrpc.HTTPResponse` with a `4xx` code, and therefore written to the writer by `respondFromHTTPGRPCResponse()`. After https://github.com/grafana/dskit/pull/457, responses with `4xx` status code are considered erroneous, and returned as errors created by the `httpgrpc` package, and therefore convertible to a `httpgrpc.HTTPResponse` with a `4xx` code. Therefore, `respondFromError()` successfully converts and writes those errors to the writer.

#### `pkg/alertmanager/Distributor.doQuorum()` ####
_Under investigation_

**Comment by duricanikolic** [comment]
## Analysis of usage of `httpgrpc`'s `client.Handle()` ##

### `RemoteQuerier.Read()` ###
**TL;DR**
This is the difference between the 2 approaches:
| returned error<br>(code, message) | BEFORE<br>(code, message)                                  | AFTER<br>(code, message) |
|-----------------------------------|------------------------------------------------------------|--------------------------|
| 4xx, this is an error             | 4xx, unexpected response status code 4xx: this is an error | 4xx, this is an error    |

**Details**

The relevant part of `RemoteQuerier.Read()` is the following:
```
func (q *RemoteQuerier) Read(...) {
	resp, err := q.client.Handle(ctx, &req)
	if err != nil {
		level.Warn(log).Log("msg", "failed to perform remote read", "err", err, "qs", query)
		return nil, err
	}
	if resp.Code/100 != 2 {
		return nil, httpgrpc.Errorf(int(resp.Code), "unexpected response status code %d: %s", resp.Code, string(resp.Body))
	}
}
```
Namely, in case of an error returned by `client.Handle()`, the error is logged and returned. Similarly, if `client.Handle()` returns no error, but the returned `httpgrpc.HTTPResponse` has a `4xx` status code, the response is converted to an error with the same status code, and with the message "_unexpected response status code `status code`: `body`_", where `status code` and `body` are the status code and the body of the response, and returned.
Before https://github.com/grafana/dskit/pull/457, responses with `4xx` status code were not considered erroneous, and were therefore converted to `httpgrpc` errors with the message pattern shown above and returned. After https://github.com/grafana/dskit/pull/457, errors with `4xx` status code are considered erroneous, and are immediately returned. 

### `RemoteQuerier.sendRequest()` ###
_Under investigation_

**Comment by pracucci** [comment]
> Analysis of usage of httpgrpc's client.Handle()
> RemoteQuerier.Read()

There's a small difference in logging. After the changes, a 4xx will be logged as `failed to perform remote read` which will generate noise for 4xx errors. I suggest to keep that error only for 5xx or in case of other non-httpgrpc errors.

> RemoteQuerier.sendRequest()

I agree we should backoff only on 5xx errors.

However, there's another difference in the caller `RemoteQuerier.query()`:  previously we were not logging `failed to remotely evaluate query expression` for 4xx, but after the change we will. I think we shouldn't log the warning for 4xx, given we expect not being an actionable error for a Mimir cluster operator.

**Comment by duricanikolic** [comment]
Hi @pracucci, 
Do you think we can close this ticket? It was intended to fix the distributor behaviour, but it was actually applied to all Mimir components using diskit's `httpgrpc/server`.

The [PR](https://github.com/grafana/backend-enterprise/pull/5737) enabling this feature in non-Mimir components is still open because there are some concerns about it. But it wasn't the goal of this issue anyway.

**Comment by pracucci** [comment]
> Do you think we can close this ticket?

I think so, thanks!
